### ISSUE TITLE
Color wrong when opening PNGs with gAMA chunk

### ISSUE THREAD

**Issue by ydahhrk** [mypaint/mypaint#129]
This is supposed to be a valid PNG (according to Gimp):

![Valid PNG](https://dl.dropboxusercontent.com/u/95836775/MyPaint/000.png)

I painted the central blob with color #BBDDAA.

If you open the image with Gimp or Krita, pick the green color and have a look at the hex value, you will see "#BBDDAA". I also tested it on a couple of closed source drawing programs, and they also report "#BBDDAA".

If you do the same with MyPaint, you get color "#BCDEAB" in the "Color Details" window.

There seems to be a problem with MyPaint's PNG loader. This also affects ORA images since they use PNG internally.

Not all PNG images I've seen suffer from this problem:

- PNG images saved by MyPaint are loaded by MyPaint correctly.
- ORA images saved by MyPaint are loaded by MyPaint correctly.
- PNG images saved by Gimp are loaded by MyPaint correctly.
- ORA images saved by Gimp are loaded by MyPaint **incorrectly**.

(That sample image was outputted by Gimp's ORA exporter; I unzipped the ORA to get the PNG).

The corruption is accumulative: MyPaint will save color #BCDEAB, since that is what it sees. Gimp will then save #BCDEAB, but MyPaint will again load it as something else (such as #BDDFAC). After several saves, the corruption is very noticeable:

![Start](https://dl.dropboxusercontent.com/u/95836775/MyPaint/start.png)

![End](https://dl.dropboxusercontent.com/u/95836775/MyPaint/end.png)

I first saw this on Ubuntu 12.04's MyPaint version. I just tested it in the master Github version and it's still a problem.

D:

**Comment by jy2wong** [comment]
I've verified the default gAMA and cHRM values against http://www.libpng.org/pub/png/spec/1.2/PNG-Chunks.html#C.sRGB . It's not clear from the standard whether this is exactly equivalent to using sRGB, or just values that are close enough. It's also not clear to me whether these are sensible defaults, or whether it's sensible to have defaults at all.

As far as I can tell, there's nothing in the specification about what you're supposed to do with just a gAMA chunk. It just says that "the absence of a cHRM chunk indicates that the image's primary colors are device-dependent."

If you comment out the following block and then remove the word else from the next 3-line block (that is, you completely ignore the gAMA chunk and default cHRM values and pretend the file uses sRGB), the colour does come out as #BBDDAA. Interesting.
https://github.com/mypaint/mypaint/blob/e7234db7e1aacdc25e1b387422e7d30868b47ae7/lib/fastpng.hpp#L405-428

I also noticed that when you open the green circle in GIMP, then check image properties with alt+return, the colour profile tab says 'sRGB built-in Default RGB working space'. So I wondered...is GIMP writing the sRGB chunk correctly?

Looking at plug-ins/common/file-png.c in the GIMP source code, it appears that no sRGB chunk is ever written. Also looking around the internet, it seems that GIMP doesn't write sRGB colour profiles, and assumes that no colour profile means sRGB.

http://www.freedesktop.org/wiki/Specifications/OpenRaster/Draft/PngDataRequirements/
However, the OpenRaster specification says that "Applications writing OpenRaster files should write appropriate colour management chunks to each PNG file in the wrapper."

I interpret this as "if the PNG uses sRGB, write an sRGB chunk".

tl;dr I think the correct long-term technical decision is to have the GIMP plugin explicitly write an sRGB chunk.

**Comment by ydahhrk** [comment]
> If you comment out the following block and then remove the word else from the next 3-line block (that is, you completely ignore the gAMA chunk and default cHRM values and pretend the file uses sRGB), the colour does come out as #BBDDAA. Interesting.

In other words, MyPaint is loading the file correctly, but everyone else is ignoring gAMA?

(A friend of mine just tested the blob image in Photoshop/Illustrator/Flash CC, and it's _also_ coming up as BBDDAA.)

**Comment by iirelu** [comment]
So we're doing it right and everyone else is doing it wrong. Sigh.

Unfortunately we can't just expect everyone else to suddenly adhere to the standards so I guess *we* just have to do things wrong as well.

**Comment by ydahhrk** [comment]
Hmmm... Something doesn't feel right.

I have two _polarizing_ opinions. I think the second one is a redder alarm, but I'm going to state the both of them for completeness sake:

-------------------------------------------

**Opinion 1**

> It's not clear from the standard whether this is exactly equivalent to using sRGB, or just values that are close enough.

I see you already figured this out, but just to make things clear:

> the exact chromaticity of the RGB data in a PNG image can be specified via the cHRM chunk, allowing more accurate color matching than gamma correction alone will provide. If the RGB data conforms to the sRGB specification [sRGB], this can be indicated with the sRGB chunk, enabling **even more accurate** reproduction.
> (http://www.libpng.org/pub/png/spec/1.2/PNG-DataRep.html#DR.Gamma-correction)

I guess gAMA plus optional cHRM equals "mediocre" sRGB? This being the case, I can sort of understand why would people utterly ignore gAMA if they're assuming no colour profile means sRGB.

/* Another reason would maybe be the implicit tendency towards deprecating the gAMA chunk? https://hsivonen.fi/png-gamma/. */

It seems like gAMA + cHRM is a fallback replacement when applications do not support sRGB, but it also seems like "do not support sRGB" is a thing of the past now (see the next link).

> it seems that GIMP (...) assumes that no colour profile means sRGB.

The more I read about these topics, the more it seems like that is actually very fair. See, for example, https://en.wikipedia.org/wiki/Talk:SRGB#Neutrality.

> So we're doing it right and everyone else is doing it wrong. Sigh.

Maybe it's not so much that people "does it wrong", but more like the standard is outdated for practical purposes?

To wrap this opinion up:

Regarding the pull request: I might be just a raving lunatic, but wouldn't it be more correct to _always_ ignore the gAMA chunk, instead of only when it's close to sRGB?

As in, it's still incorrect according to the standard, but if everyone is assuming no colorspace means sRGB, and gAMA is a fallback option when sRGB is not present, then it sounds like nobody is actually using gAMA.

Also, they are tools intended for the same purpose. Their values are _supposed_ to match.

-------------------------------------------

**Opinion 2**

And something is really honestly bothering me about this.

I mean @jy2wong noticed that removing that else seems to fix the problem, but I feel like we're jumping to conclusions.

I understand that the reason why MyPaint is presently picking (**and saving**) value #BCDEAB is because it's applying/committing the gamma value on the final pixels. Am I wrong?

Doesn't this defeat the whole purpose of gamma?

As I read about this, I get the impression that the whole point of gamma is that the bare color values by themselves are, in a way, tied to the technology that generated them and need to be transformed in order to reproduce the same feeling when perceived by technology with different properties.

> Ordinarily, the editing application would simply write a gamma value that corresponds to the artist's display system. But if the image in question originated on another system, some editors will actually preserve its gamma setting by using a decoding_exponent for all manipulations on the artist's system--just as a normal viewer would. Thus the artist sees an image displayed in her own ``gamma space,'' but the underlying image samples actually remain in the gamma space of the original system.
> (http://libpng.org/pub/png/book/chapter10.html)

If that is not the case... if MyPaint is commiting the file's gamma, and also its own... isn't that the source of the corruption?

Do you guys write a gAMA chunk? Because if not, it would be like,

- Gimp: OK, the image has no gAMA chunk but the user is asking me to write one... alright, I'll do that.
- MyPaint: Alright... the image has a gAMA chunk but whatever... I'm going to apply it on the pixels. Oh, the user is asking me to write the file... I don't have any gAMA thing in memory so I won't include it. [Gamma collapsed to pixels, pixels corrupted]
- Gimp: OK, the image has no gAMA chunk but the user is asking me to write one... alright, I'll do that.

ad infinitum.

But even if MyPaint remembers and writes the gAMA chunk this might still be a problem because it still seems like it's effectively applying it on the pixels themselves.

Am I making sense?

**Comment by jy2wong** [comment]
I exported a fresh PNG in mypaint and looked at the hexdump with `hexdump -c foo.png | less`. I don't see a gAMA chunk (or an sRGB chunk, for that matter).

**Comment by 0ion9** [comment]
* MyPaint: Alright... the image has a gAMA chunk but whatever... I'm going to apply it on the pixels. Oh, the user is asking me to write the file... I don't have any gAMA thing in memory so I won't include it. 

IMO this is the correct behaviour, at least until we support such things as customizable working colorspace, because we *are* working in sRGB. It's closely analogous to GIMP's 'this image has a ICC profile attached; want to convert to the default profile(sRGB)?'; a necessary measure to make the images we throw around inside the app consistent with each other. One way or another, whether we decide to apply gAMA or not, we cannot avoid this 'corruption' without more complete internal color management than we currently have.

That said, if this policy breaks on a majority of cases, we are pretty much obliged to do the wrong thing, at least until we have full internal color management.

(A further wrinkle that IIRC no-one's pointed out yet: On Macs, Gamma=1.8 is the appropriate coefficient for sRGB, while Gamma=2.2 is correct for Windows and Linux. Any attempt to identify sRGB by gamma/chromaticity needs to account for platform.)

**Comment by achadwick** [comment]
> (A further wrinkle that IIRC no-one's pointed out yet: On Macs, Gamma=1.8 is the appropriate coefficient for sRGB, while Gamma=2.2 is correct for Windows and Linux. Any attempt to identify sRGB by gamma/chromaticity needs to account for platform.)

No. sRGB is a standard colour space, and it always uses the same primaries and the same (rather unusual) gamma curve: https://en.wikipedia.org/wiki/SRGB no matter what platform.

**Comment by achadwick** [comment]
Looks like we have sRGB saving turned off: when we transitioned from no colour processing on input to respecting PNGs' saved colourspace info (and converting them to an assumed sRGB), I thought that we might use the absence of such a flag as an indication that the PNG to be loaded was a legacy one. The default is set in lib.pixbufsurface.save_as_png(), and it isn't overridden anywhere.

It's more

* MyPaint: Alright... the image has a gAMA chunk but whatever... I'm going to apply it on the pixels to convert it to my internal 15+1-bit fixed point representation and assumed sRGB colour space. Oh, the user is asking me to write the file... I'll save it old-style so I might recognise that later on.

I think now that we should save our assumption. My thinking:

Long term, MyPaint needs to be aware of display colour spaces (defaulting to sRGB) and either

1. Convert to the display space on load  → render the pixels unchanged → and save layer PNGs with the internal data unchanged with the colour space tagged.
2. Load unchanged → convert to the display space for rendering → save unchanged.

The first option is closer to what we do at the moment, and if we do it properly requires saving with the `sRGB` chunk (and ideally with a `gAMA` which approximates sRGB, and sRGB's primaries in `cHRM`).
The second option is a lot closer to OpenRaster's intent: the spec has always stressed pass-through of unrecognized layer data as a primary concern. I favour this, and want to work towards this model.

See also #6 . I really want linear light compositing too if we can, but this would be a *major* colour consistency issue.

I note that we now support version numbering of saved files, which may help a lot for this sort of plan. I always want stable releases of MyPaint to be able to load files saved with older stable releases sensibly.

Option 2 is more computationally expensive for rendering. However we now cache rendered tile data in an LRU cache, which will help.

**Comment by achadwick** [comment]
@jy2wong 
> I exported a fresh PNG in mypaint and looked at the hexdump with hexdump -c foo.png | less. I don't see a gAMA chunk (or an sRGB chunk, for that matter).

Thanks for the reminder. Can people with an interest in interoperability with GIMP please test https://github.com/achadwick/mypaint/tree/save-png-sRGB-chunks to ensure that it doesn't introduce a regression when saving files from MyPaint then loading them into GIMP.

Using `pngchunks` on the original #BBDDAA green-blob file attached above, as resaved by MyPaint with 40a70b1 shows an `sRGB` chunk and the `gAMA`/`cHRM` approximations needed for backwards compatibility. In other words, PNGs saved from MyPaint begin to reflect the embarrassing internal assumptions we've been making all along.

As mentioned above and in the comment on 40a70b1, I don't think absence of colour management info on loaded PNGs can ever be used as an indication that the PNG was saved by an earlier MyPaint. I'll be ripping all "legacy png" detection out as meaningless cruft if achadwick/mypaint/tree/save-png-sRGB-chunks plays nice on saving, so please test.

**Comment by achadwick** [comment]
Apologies for the forced push.

Perhaps the answer to the legacy vs. nonlegacy question is this, also on the branch now:

![toggle](https://cloud.githubusercontent.com/assets/61299/5986516/4589ad1a-a8f3-11e4-85e9-c11a65814b6b.png)

on by default, it makes MyPaint convert PNGs tagged with colour management info to the internal sRGB space, and save with sRGB chunks. It's a very basic step towards proper colour management, but one I think we need to take.

**Comment by iirelu** [comment]
If that description is for more than just debugging it definitely needs to have a more human readable description. Our users are artists, after all.

**Comment by achadwick** [comment]
Pro artists need to know about colour transforms being applied. Happy to shorten it though if we can find a really brief way of summing it up.

It's on the Load & Save tab because it only governs stuff that happens during those ops. So what about a radio button:

    (*) Convert files to sRGB when loading, tag them as sRGB when saving
    ( ) No colour space conversions or tagging

Radio buttons are often more self-explanatory.

**Comment by ydahhrk** [comment]
> I favour this, and want to work towards this model.

I second this, but it seems like you are actually coding the other model. I take that you want to use the first one as a temporary short-term solution, and eventually implement the second model in the future?

> Can people with an interest in interoperability with GIMP please test https://github.com/achadwick/mypaint/tree/save-png-sRGB-chunks to ensure that it doesn't introduce a regression when saving files from MyPaint then loading them into GIMP.

My results are baffling, but maybe I got the message completely backwards, so I'm going to be factual first. Here's what I did:

- Using MyPaint (2e6e5081baf973b682b93cc6f9bf44797b49ebd9), I created two files. One with "Assume display space is sRGB" active (`sRGB`, `gAMA` and `cHRM`), the other one with the option disabled (no optional chunks).
- Using the GIMP, I created two more files. One with `gAMA`, the other without `gAMA`.
- Using Krita, I created two files. One with sRGB profile (`iCCP` & `pHYs`), the other one without (`sRGB` & `pHYs`).

(Krita doesn't actually seem to introduce any noise, but I wanted another opinion.)

The files are the folderless PNGs in [this zip](https://dl.dropboxusercontent.com/u/95836775/MyPaint/tests1.zip). Eg. `mypaint assume-sRGB.png` is the file saved w/MyPaint with "Assume display space is sRGB" active.

Then I opened each file and saved it as every other combination. These are the subfolders. Eg. `krita sRGB-profile/gimp gamma.png` is `gimp-gamma.png` resaved using Krita with the sRGB profile.

Then I opened each file (including the root ones) using each program (where "MyPaint assuming sRGB" is considered one program, "MyPaint not assuming sRGB" is considered another program). These are the results:

Every combination reports "#BBDDAA", except the following:

- "mypaint assume-sRGB/gimp gamma.png" with ANY program.
- "gimp gamma.png" using MyPaint in mode "Assume sRGB".
- Every file under the "gimp gamma" folder using MyPaint in mode "Assume sRGB".

Is this what you expected? (Both saving `gAMA` in the GIMP and turning on "Assume sRGB" in MyPaint seem to be poisonous)

**Comment by iirelu** [comment]
@achadwick I'll make an other issue (probably) for the specifics of the UI, this one's cramped enough as it is.

**Comment by ydahhrk** [comment]
PS. Keep in mind that a workaround _exists_, so even if this "proper colour management" thing would take long, you don't HAVE to spend too much effort trying to awkwardly patch this side-effect.

But if you're willing to keep digging for some sort of trade-off, I am.

Is the relevant code for the "proper colour management" too vast for a noob like me?

**Comment by achadwick** [comment]
Okay, can people please test master now? We have a specific workaround for GIMP's gamma-2.2 guesstimate, and do no colour transforms in that case *or* when the file says sRGB. We're pretending that the display is two things.

I've made sure that [pngsuite's gamma tests](http://www.schaik.com/pngsuite/#gamma) and [libpng's cHRM tests](http://libpng.org/pub/png/png-colortest.html) all still look OK.

The prefs dialog now looks like 
![dialognow](https://cloud.githubusercontent.com/assets/61299/6091350/e051e932-ae9f-11e4-9d8b-607c971fe881.png)
and provides another workaround for issues like this.

**Comment by ydahhrk** [comment]
[tests2.zip](https://dl.dropboxusercontent.com/u/95836775/MyPaint/tests2.zip)

It looks almost "too" good :p. I was sort of expecting (and willing to accept) some color changes when I mixed configurations recklessly, such as saving w/sRGB then dropping the chunks in the GIMP and then reopening with or without conversion, but I guess you figured it out really well.
(is this expected?)

Also, do you think there's value in testing with actual image edits?
I'm just opening and saving with the different configurations. Maybe if I actually edit before saving, and sample both unedited and edited pixels, interesting stuff might happen?

If this doesn't set off any more alarms, I believe this issue is ready to be closed.

**Comment by achadwick** [comment]
Heh. looked like I jumped the gun on closing this one with the commit message. Thanks for testing it out though!

I'll leave this bug closed unless someone has any objections?

--------------------

Incidentally, my colorimeter order arrived yesterday. Guess I'll be playing with that after the upcoming release!
